### PR TITLE
Rework create project endpoint

### PR DESCRIPTION
**Pull request by djino97**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/409

### CODE REVIEW

File: src/ProjectService.Models.Dto/Enums/UserRoleType.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.ProjectService.Models.Dto.Enums
+{
+    public enum UserRoleType
+    {
+        Admin
```
Comment (KostinP) [inline src/ProjectService.Models.Dto/Enums/UserRoleType.cs:5]: TeamLead?

File: src/ProjectService.Data/ProjectRepository.cs [hunk 2]
```diff
@@ -43,12 +43,10 @@ public IEnumerable<DbProjectUser> GetProjectUsers(Guid projectId, bool showNotAc
             return provider.ProjectsUsers.Include(u => u.Role).Where(predicate).ToList();
         }
 
-        public Guid CreateNewProject(DbProject newProject)
+        public void CreateNewProject(DbProject newProject)
```
Comment (KostinP) [inline src/ProjectService.Data/ProjectRepository.cs:46]: Why?

File: src/ProjectService.Mappers/ModelsMappers/ProjectUserMapper.cs [hunk 3]
```diff
@@ -14,16 +14,16 @@ namespace LT.DigitalOffice.ProjectService.Mappers.ModelsMappers
...
         public ProjectUserMapper(
             ILogger<ProjectUserMapper> logger,
-            IRoleMapper roleMapper,
+            //IRoleMapper roleMapper,
             IRequestClient<IGetUserDataRequest> requestClient)
         {
             _logger = logger;
-            _roleMapper = roleMapper;
+           // _roleMapper = roleMapper;
             _requestClient = requestClient;
```
Comment (KostinP) [inline src/ProjectService.Mappers/ModelsMappers/ProjectUserMapper.cs:27]: //?

File: src/ProjectService.Mappers/RequestsMappers/DbProjectMapper.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                Id = projectId,
+                AuthorId = authorId,
+                Name = request.Name,
+                Status = (int)request.Status,
+                ShortName = request.ShortName,
+                Description = request.Description,
+                ShortDescription = request.ShortDescription,
+                DepartmentId = request.DepartmentId,
+                CreatedAt = DateTime.UtcNow,
+                Users = request.Users?
```
Comment (KostinP) [inline src/ProjectService.Mappers/RequestsMappers/DbProjectMapper.cs:32]: Can be null?

File: src/ProjectService.Mappers/ResponsesMappers/ProjectResponseMapper.cs [hunk 5]
```diff
@@ -17,14 +17,14 @@ public Project Map(DbProject value)
...
+                /*Name = value.Name,
                 ShortName = value.ShortName,
                 CreatedAt = value.CreatedAt,
-                ClosedAt = value.ClosedAt,
+                //ClosedAt = value.ClosedAt,
                 DepartmentId = value.DepartmentId,
-                ClosedReason = value.ClosedReason,
-                Description = value.Description
+               // ClosedReason = value.ClosedReason,
+                Description = value.Description*/
```
Comment (KostinP) [inline src/ProjectService.Mappers/ResponsesMappers/ProjectResponseMapper.cs:27]: //?

File: src/ProjectService.Models.Dto/Models/ProjectUser.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿using LT.DigitalOffice.ProjectService.Models.Dto.Enums;
+using System;
+
+namespace LT.DigitalOffice.ProjectService.Models.Dto.RequestsModels
+{
+    public class ProjectUser
```
Comment (KostinP) [inline src/ProjectService.Models.Dto/Models/ProjectUser.cs:6]: ProjectUserInfo

File: test/ProjectService.Business.UnitTests/Commands/GetProjectByIdCommandTests.cs [hunk 7]
```diff
@@ -152,7 +151,7 @@ public void ShouldReturnProjectInfo()
             var result = command.Execute(projectId, SHOW_NOT_ACTIVE_USERS).Result;
 
             Assert.AreEqual(expectedResult.Project.Id, result.Project.Id);
-            Assert.AreEqual(expectedResult.Project.Name, result.Project.Name);
+           // Assert.AreEqual(expectedResult.Project.Name, result.Project.Name);
```
Comment (KostinP) [inline test/ProjectService.Business.UnitTests/Commands/GetProjectByIdCommandTests.cs:154]: //?

File: src/ProjectService.Models.Dto/Models/ProjectUser.cs (hunk 6 above)
Comment (djino97) [reply src/ProjectService.Models.Dto/Models/ProjectUser.cs:6]: This is not info of project user

File: src/ProjectService.Data/ProjectRepository.cs (hunk 2 above)
Comment (djino97) [reply src/ProjectService.Data/ProjectRepository.cs:46]: What?

File: src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs [hunk 8]
```diff
@@ -0,0 +1,81 @@
+﻿using LT.DigitalOffice.ProjectService.Models.Db;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Migrations;
+using System;
+
+namespace LT.DigitalOffice.ProjectService.Data.Provider.MsSql.Ef.Migrations
+{
+    [DbContext(typeof(ProjectServiceDbContext))]
+    [Migration("20210419235819_InitialCreate")]
```
Comment (MariyaSin) [inline src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs:9]: Are you sure what it's correct migration name?
I think it sould be "20210419235819_UpdateProjectsAndProjectsUsersTables"

Comment (djino97) [reply src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs]: Fixed

File: src/ProjectService.Models.Db/DbProject.cs [hunk 9]
```diff
@@ -40,6 +40,14 @@ public void Configure(EntityTypeBuilder<DbProject> builder)
             builder
                 .HasKey(p => p.Id);
 
+            builder
+                .Property(P => P.Name)
+                .IsRequired();
```
Comment (MariyaSin) [inline src/ProjectService.Models.Db/DbProject.cs:45]: in 55 line already exist this  condition

File: src/ProjectService.Mappers/ModelsMappers/ProjectUserMapper.cs (hunk 3 above)
Comment (djino97) [reply src/ProjectService.Mappers/ModelsMappers/ProjectUserMapper.cs]: Fixed

File: src/ProjectService.Mappers/ResponsesMappers/ProjectResponseMapper.cs (hunk 5 above)
Comment (djino97) [reply src/ProjectService.Mappers/ResponsesMappers/ProjectResponseMapper.cs]: Fixed

File: src/ProjectService.Models.Db/DbProject.cs [hunk 10]
```diff
@@ -40,6 +40,14 @@ public void Configure(EntityTypeBuilder<DbProject> builder)
...
+
+            builder
+                .Property(p => p.ShortName)
+                .IsRequired();
+
```
Comment (MariyaSin) [inline src/ProjectService.Models.Db/DbProject.cs:50]: I think you need to add conditions to DepartmentId and Status.

Comment (djino97) [reply src/ProjectService.Models.Db/DbProject.cs:50]: this are structure

File: src/ProjectService.Models.Db/DbProject.cs (hunk 9 above)
Comment (djino97) [reply src/ProjectService.Models.Db/DbProject.cs:45]: Fixed

File: test/ProjectService.Business.UnitTests/Commands/GetProjectByIdCommandTests.cs (hunk 7 above)
Comment (djino97) [reply test/ProjectService.Business.UnitTests/Commands/GetProjectByIdCommandTests.cs]: fixed

File: src/ProjectService.Models.Dto/Models/ProjectUser.cs (hunk 6 above)
Comment (djino97) [reply src/ProjectService.Models.Dto/Models/ProjectUser.cs]: Fixed

File: src/ProjectService.Data/ProjectRepository.cs (hunk 2 above)
Comment (KostinP) [reply src/ProjectService.Data/ProjectRepository.cs:46]: Response: base project info

Comment (djino97) [reply src/ProjectService.Data/ProjectRepository.cs:46]: Why? I do it in command

**Review by MariyaSin — APPROVED** [review]

File: src/ProjectService.Mappers/RequestsMappers/DbProjectMapper.cs (hunk 4 above)
Comment (djino97) [reply src/ProjectService.Mappers/RequestsMappers/DbProjectMapper.cs:32]: Yea

File: src/ProjectService.Data/ProjectRepository.cs (hunk 2 above)
Comment (KostinP) [reply src/ProjectService.Data/ProjectRepository.cs:46]: Because it is written in the technical description

File: src/ProjectService.Models.Dto/Enums/UserRoleType.cs (hunk 1 above)
Comment (djino97) [reply src/ProjectService.Models.Dto/Enums/UserRoleType.cs]: fixed

**Review by KostinP — APPROVED** [review]

File: src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            UpdateProjectsTable(migrationBuilder);
+            UpdateProjectUsersTable(migrationBuilder);
+
+            migrationBuilder.DropTable("Roles");
+        }
+
+        protected override void Down(MigrationBuilder migrationBuilder)
```
Comment (BelehovEgor) [inline src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs:79]: why did you add this method, and didn't write anything?

File: src/ProjectService.Data/Interfaces/IProjectRepository.cs [hunk 12]
```diff
@@ -23,14 +23,13 @@ public interface IProjectRepository
         /// Adds new project to the database. Returns the id of the added project.
         /// </summary>
         /// <param name="item">Project to add.</param>
-        /// <returns>Id of the added project.</returns>
-        Guid CreateNewProject(DbProject item);
+        void CreateNewProject(DbProject item);
```
Comment (BelehovEgor) [inline src/ProjectService.Data/Interfaces/IProjectRepository.cs:26]: why don't we return the id now?

File: src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+using LT.DigitalOffice.Kernel.Attributes;
+using LT.DigitalOffice.ProjectService.Models.Db;
+using LT.DigitalOffice.ProjectService.Models.Dto.Models;
+
+namespace LT.DigitalOffice.ProjectService.Mappers.ModelsMappers.Interfaces
+{
+    [AutoInject]
+    public interface IProjectInfoMapper
+    {
+        ProjectInfo Map(DbProject dbProject, IGetDepartmentResponse departmet);
```
Comment (BelehovEgor) [inline src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs:11]: I would only send the name of the department and not the entire structure

File: src/ProjectService.Models.Dto/Enums/ProjectStatusType.cs [hunk 14]
```diff
@@ -1,10 +1,12 @@
 ﻿namespace LT.DigitalOffice.ProjectService.Models.Dto.Enums
 {
-    public enum ProjectClosedReason
+    public enum ProjectStatusType
```
Comment (BelehovEgor) [inline src/ProjectService.Models.Dto/Enums/ProjectStatusType.cs:3]: I still think. that we need only 3 types, but it is not critical

File: src/ProjectService.Models.Dto/Enums/UserRoleType.cs [hunk 15]
```diff
@@ -0,0 +1,7 @@
+﻿namespace LT.DigitalOffice.ProjectService.Models.Dto.Enums
+{
+    public enum UserRoleType
```
Comment (BelehovEgor) [inline src/ProjectService.Models.Dto/Enums/UserRoleType.cs:3]: only one role?

File: src/ProjectService.Models.Dto/Models/ProjectInfo.cs [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+namespace LT.DigitalOffice.ProjectService.Models.Dto.Models
+{
+    public class ProjectInfo
+    {
+        public Guid Id { get; set; }
+        public Guid AuthorId { get; set; }
+        public ProjectStatusType Status { get; set; }
+        public string Name { get; set; }
+        public string ShortName { get; set; }
+        public string Description { get; set; }
```
Comment (BelehovEgor) [inline src/ProjectService.Models.Dto/Models/ProjectInfo.cs:13]: i think we don't need it in base info

File: src/ProjectService.Validation/ProjectUserValidator.cs [hunk 17]
```diff
@@ -8,19 +8,14 @@ public class ProjectUserValidator : AbstractValidator<ProjectUserRequest>, IProj
...
         {
-            RuleFor(pu => pu.User)
+            RuleFor(pu => pu)
                 .Must(u => u != null)
                 .WithMessage("The request must contain user data")
                 .DependentRules(() =>
                 {
-                    RuleFor(pu => pu.User.Id)
+                    RuleFor(pu => pu.UserId)
                     .NotEmpty()
```
Comment (BelehovEgor) [inline src/ProjectService.Validation/ProjectUserValidator.cs:17]: add tabulation

File: src/ProjectService.Models.Dto/Enums/UserRoleType.cs (hunk 15 above)
Comment (djino97) [reply src/ProjectService.Models.Dto/Enums/UserRoleType.cs:3]: to date, yes

File: src/ProjectService.Models.Dto/Enums/ProjectStatusType.cs (hunk 14 above)
Comment (djino97) [reply src/ProjectService.Models.Dto/Enums/ProjectStatusType.cs:3]: yes, not critical

File: src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs (hunk 11 above)
Comment (djino97) [reply src/ProjectService.Data.Provider.MsSql.Ef/Migrations/20210419235819_UpdateProjectsAndProjectsUsersTables.cs]: Fixed

File: src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs (hunk 13 above)
Comment (djino97) [reply src/ProjectService.Mappers/ModelsMappers/Interfaces/IProjectInfoMapper.cs]: Fixed

File: src/ProjectService.Validation/ProjectUserValidator.cs (hunk 17 above)
Comment (djino97) [reply src/ProjectService.Validation/ProjectUserValidator.cs]: fixed

**Review by BelehovEgor — APPROVED** [review]

**Review by TIove — APPROVED** [review]
